### ISSUE TITLE
Out of bounds vector access

### ISSUE THREAD

**Issue by jamesjer** [libsemigroups/libsemigroups#391]
I attempted to build libsemigroups 2.3.0 for Fedora.  The Fedora build flags include `-D_GLIBCXX_ASSERTIONS`, which I highly recommend for testing builds.  One test failed:
```
-------------------------------------------------------------------------------
Sims1 000: fp example 1
-------------------------------------------------------------------------------
tests/test-sims1.cpp:77
...............................................................................

tests/test-sims1.cpp:77: FAILED:
due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal

Sims1 000: fp example 1 . . .                                                  
===============================================================================
test cases:     869 |     868 passed | 1 failed
assertions: 1162775 | 1162774 passed | 1 failed

/usr/include/c++/12/bits/stl_vector.h:1123: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = libsemigroups::Sims1<unsigned int>::PendingDef; _Alloc = std::allocator<libsemigroups::Sims1<unsigned int>::PendingDef>; reference = libsemigroups::Sims1<unsigned int>::PendingDef&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
make: *** [Makefile:5233: check] Aborted (core dumped)
```

Weirdly, I cannot get this to happen with GDB, but it happens every time when running the test from the command line.  Valgrind, on the other hand, provided a backtrace:
```
==68== Process terminating with default action of signal 6 (SIGABRT)
==68==    at 0x4D2D2EC: __pthread_kill_implementation (pthread_kill.c:44)
==68==    by 0x4CDB005: raise (raise.c:26)
==68==    by 0x4CC47FB: abort (abort.c:79)
==68==    by 0x4A46EAF: std::__glibcxx_assert_fail(char const*, int, char const*, char const*) (debug.cc:60)
==68==    by 0x95A7EE: UnknownInlinedFun (stl_vector.h:1123)
==68==    by 0x95A7EE: UnknownInlinedFun (stl_vector.h:1121)
==68==    by 0x95A7EE: UnknownInlinedFun (sims1.tpp:707)
==68==    by 0x95A7EE: UnknownInlinedFun (sims1.tpp:719)
==68==    by 0x95A7EE: UnknownInlinedFun (sims1.tpp:790)
==68==    by 0x95A7EE: libsemigroups::Sims1<unsigned int>::thread_runner::worker_thread(unsigned int, std::function<bool (libsemigroups::ActionDigraph<unsigned int> const&)>) (sims1.tpp:747)
==68==    by 0x956113: UnknownInlinedFun (invoke.h:74)
==68==    by 0x956113: UnknownInlinedFun (invoke.h:96)
==68==    by 0x956113: UnknownInlinedFun (std_thread.h:252)
==68==    by 0x956113: UnknownInlinedFun (std_thread.h:259)
==68==    by 0x956113: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (libsemigroups::Sims1<unsigned int>::thread_runner::*)(unsigned int, std::function<bool (libsemigroups::ActionDigraph<unsigned int> const&)>), libsemigroups::Sims1<unsigned int>::thread_runner*, unsigned long, std::reference_wrapper<std::function<bool (libsemigroups::ActionDigraph<unsigned int> const&)> > > > >::_M_run() (std_thread.h:210)
==68==    by 0x4A4AC32: execute_native_thread_routine (thread.cc:82)
==68==    by 0x4D2B5B4: start_thread (pthread_create.c:442)
==68==    by 0x4DABF03: clone (clone.S:100)
```

Looking at sims1.tpp line 707 shows that we are on the last line of this loop:
```
      for (size_t i = 0; i < that._pending.size(); i += 2) {
        this->_pending.push_back(std::move(that._pending[i]));
        that._pending[i / 2] = std::move(that._pending[i + 1]);
      }
```

Therefore, `that._pending.size()` must be an odd number, so the `std::move` call is trying to move an element beyond the end of the vector.

**Comment by james-d-mitchell** [comment]
Thanks for the report @jamesjer, I very much appreciate it. This should be easy to fix, I'll try to do it now.

**Comment by james-d-mitchell** [comment]
Hopefully this was resolved in v2.3.1

**Comment by jamesjer** [comment]
I can confirm this was resolved in 2.3.1.  Thank you for the quick response!
